// odata-v4-ng/esm5/lib/odata-response/csdl/csdl-annotation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
var CsdlAnnotable = /** @class */ (function () {
    function CsdlAnnotable(annotationList) {
        this.annotationList = annotationList;
    }
    return CsdlAnnotable;
}());
export { CsdlAnnotable };
if (false) {
    /** @type {?} */
    CsdlAnnotable.prototype.annotationList;
}
var CsdlAnnotations = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlAnnotations, _super);
    function CsdlAnnotations(target, annotationList, qualifier) {
        var _this = _super.call(this, annotationList) || this;
        _this.target = target;
        _this.qualifier = qualifier;
        return _this;
    }
    return CsdlAnnotations;
}(CsdlAnnotable));
export { CsdlAnnotations };
if (false) {
    /** @type {?} */
    CsdlAnnotations.prototype.target;
    /** @type {?} */
    CsdlAnnotations.prototype.qualifier;
}
var CsdlAnnotation = /** @class */ (function () {
    function CsdlAnnotation(term, qualifier) {
        this.term = term;
        this.qualifier = qualifier;
    }
    return CsdlAnnotation;
}());
export { CsdlAnnotation };
if (false) {
    /** @type {?} */
    CsdlAnnotation.prototype.term;
    /** @type {?} */
    CsdlAnnotation.prototype.qualifier;
}
var CsdlTerm = /** @class */ (function () {
    function CsdlTerm(name, type, baseTerm, defaultValue, appliesTo, nullable, maxLength, precision, scale, srid) {
        this.name = name;
        this.type = type;
        this.baseTerm = baseTerm;
        this.defaultValue = defaultValue;
        this.appliesTo = appliesTo;
        this.nullable = nullable;
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.srid = srid;
    }
    return CsdlTerm;
}());
export { CsdlTerm };
if (false) {
    /** @type {?} */
    CsdlTerm.prototype.name;
    /** @type {?} */
    CsdlTerm.prototype.type;
    /** @type {?} */
    CsdlTerm.prototype.baseTerm;
    /** @type {?} */
    CsdlTerm.prototype.defaultValue;
    /** @type {?} */
    CsdlTerm.prototype.appliesTo;
    /** @type {?} */
    CsdlTerm.prototype.nullable;
    /** @type {?} */
    CsdlTerm.prototype.maxLength;
    /** @type {?} */
    CsdlTerm.prototype.precision;
    /** @type {?} */
    CsdlTerm.prototype.scale;
    /** @type {?} */
    CsdlTerm.prototype.srid;
}
