// odata-v4-ng/esm5/lib/odata-response/csdl/csdl-reference.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CsdlAnnotable } from './csdl-annotation';
var CsdlReference = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlReference, _super);
    function CsdlReference(uri, includes, includeAnnotations, annotationList) {
        var _this = _super.call(this, annotationList) || this;
        _this.uri = uri;
        _this.includes = includes;
        _this.includeAnnotations = includeAnnotations;
        return _this;
    }
    return CsdlReference;
}(CsdlAnnotable));
export { CsdlReference };
if (false) {
    /** @type {?} */
    CsdlReference.prototype.uri;
    /** @type {?} */
    CsdlReference.prototype.includes;
    /** @type {?} */
    CsdlReference.prototype.includeAnnotations;
}
var CsdlInclude = /** @class */ (function () {
    function CsdlInclude(namespace, alias) {
        this.namespace = namespace;
        this.alias = alias;
    }
    return CsdlInclude;
}());
export { CsdlInclude };
if (false) {
    /** @type {?} */
    CsdlInclude.prototype.namespace;
    /** @type {?} */
    CsdlInclude.prototype.alias;
}
var CsdlIncludeAnnotations = /** @class */ (function () {
    function CsdlIncludeAnnotations(termNamespace, qualifier, targetNamespace) {
        this.termNamespace = termNamespace;
        this.qualifier = qualifier;
        this.targetNamespace = targetNamespace;
    }
    return CsdlIncludeAnnotations;
}());
export { CsdlIncludeAnnotations };
if (false) {
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.termNamespace;
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.qualifier;
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.targetNamespace;
}
