// odata-v4-ng/esm5/lib/odata-response/csdl/csdl-structural-property.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * @abstract
 */
var /**
 * @abstract
 */
CsdlStructuralProperty = /** @class */ (function () {
    function CsdlStructuralProperty(name, type, nullable) {
        this.name = name;
        this.type = type;
        this.nullable = nullable;
    }
    return CsdlStructuralProperty;
}());
/**
 * @abstract
 */
export { CsdlStructuralProperty };
if (false) {
    /** @type {?} */
    CsdlStructuralProperty.prototype.name;
    /** @type {?} */
    CsdlStructuralProperty.prototype.type;
    /** @type {?} */
    CsdlStructuralProperty.prototype.nullable;
}
var CsdlProperty = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlProperty, _super);
    function CsdlProperty(name, type, nullable, maxLength, precision, scale, unicode, srid, defaultValue) {
        var _this = _super.call(this, name, type, nullable) || this;
        _this.maxLength = maxLength;
        _this.precision = precision;
        _this.scale = scale;
        _this.unicode = unicode;
        _this.srid = srid;
        _this.defaultValue = defaultValue;
        return _this;
    }
    return CsdlProperty;
}(CsdlStructuralProperty));
export { CsdlProperty };
if (false) {
    /** @type {?} */
    CsdlProperty.prototype.maxLength;
    /** @type {?} */
    CsdlProperty.prototype.precision;
    /** @type {?} */
    CsdlProperty.prototype.scale;
    /** @type {?} */
    CsdlProperty.prototype.unicode;
    /** @type {?} */
    CsdlProperty.prototype.srid;
    /** @type {?} */
    CsdlProperty.prototype.defaultValue;
}
var CsdlNavigationProperty = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlNavigationProperty, _super);
    function CsdlNavigationProperty(name, type, nullable, partner, containsTarget, referentialConstraints, onDelete) {
        var _this = _super.call(this, name, type, nullable) || this;
        _this.partner = partner;
        _this.containsTarget = containsTarget;
        _this.referentialConstraints = referentialConstraints;
        _this.onDelete = onDelete;
        return _this;
    }
    return CsdlNavigationProperty;
}(CsdlStructuralProperty));
export { CsdlNavigationProperty };
if (false) {
    /** @type {?} */
    CsdlNavigationProperty.prototype.partner;
    /** @type {?} */
    CsdlNavigationProperty.prototype.containsTarget;
    /** @type {?} */
    CsdlNavigationProperty.prototype.referentialConstraints;
    /** @type {?} */
    CsdlNavigationProperty.prototype.onDelete;
}
var CsdlReferentialConstraint = /** @class */ (function () {
    function CsdlReferentialConstraint(property, referencedProperty) {
        this.property = property;
        this.referencedProperty = referencedProperty;
    }
    return CsdlReferentialConstraint;
}());
export { CsdlReferentialConstraint };
if (false) {
    /** @type {?} */
    CsdlReferentialConstraint.prototype.property;
    /** @type {?} */
    CsdlReferentialConstraint.prototype.referencedProperty;
}
var CsdlOnDelete = /** @class */ (function () {
    function CsdlOnDelete(action) {
        this.action = action;
    }
    return CsdlOnDelete;
}());
export { CsdlOnDelete };
if (false) {
    /** @type {?} */
    CsdlOnDelete.prototype.action;
}
