// odata-v4-ng/esm5/lib/odata-response/csdl/csdl-enum-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var CsdlEnumType = /** @class */ (function () {
    function CsdlEnumType(name, members, underlyingType, isFlags) {
        this.name = name;
        this.members = members;
        this.underlyingType = underlyingType;
        this.isFlags = isFlags;
    }
    return CsdlEnumType;
}());
export { CsdlEnumType };
if (false) {
    /** @type {?} */
    CsdlEnumType.prototype.name;
    /** @type {?} */
    CsdlEnumType.prototype.members;
    /** @type {?} */
    CsdlEnumType.prototype.underlyingType;
    /** @type {?} */
    CsdlEnumType.prototype.isFlags;
}
var CsdlEnumMember = /** @class */ (function () {
    function CsdlEnumMember(name, value) {
        this.name = name;
        this.value = value;
    }
    return CsdlEnumMember;
}());
export { CsdlEnumMember };
if (false) {
    /** @type {?} */
    CsdlEnumMember.prototype.name;
    /** @type {?} */
    CsdlEnumMember.prototype.value;
}
