// odata-v4-ng/esm5/lib/odata-response/csdl/csdl-structured-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
var CsdlStructuredType = /** @class */ (function () {
    function CsdlStructuredType(name, properties, navigationProperties, baseType, openType, abstract) {
        this.name = name;
        this.properties = properties;
        this.navigationProperties = navigationProperties;
        this.baseType = baseType;
        this.openType = openType;
        this.abstract = abstract;
    }
    return CsdlStructuredType;
}());
export { CsdlStructuredType };
if (false) {
    /** @type {?} */
    CsdlStructuredType.prototype.name;
    /** @type {?} */
    CsdlStructuredType.prototype.properties;
    /** @type {?} */
    CsdlStructuredType.prototype.navigationProperties;
    /** @type {?} */
    CsdlStructuredType.prototype.baseType;
    /** @type {?} */
    CsdlStructuredType.prototype.openType;
    /** @type {?} */
    CsdlStructuredType.prototype.abstract;
}
var CsdlComplexType = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlComplexType, _super);
    function CsdlComplexType(name, properties, navigationProperties, baseType, openType, abstract) {
        return _super.call(this, name, properties, navigationProperties, baseType, openType, abstract) || this;
    }
    return CsdlComplexType;
}(CsdlStructuredType));
export { CsdlComplexType };
var CsdlEntityType = /** @class */ (function (_super) {
    tslib_1.__extends(CsdlEntityType, _super);
    function CsdlEntityType(name, key, properties, navigationProperties, baseType, openType, abstract, hasStream) {
        var _this = _super.call(this, name, properties, navigationProperties, baseType, openType, abstract) || this;
        _this.key = key;
        _this.hasStream = hasStream;
        return _this;
    }
    return CsdlEntityType;
}(CsdlStructuredType));
export { CsdlEntityType };
if (false) {
    /** @type {?} */
    CsdlEntityType.prototype.key;
    /** @type {?} */
    CsdlEntityType.prototype.hasStream;
}
var CsdlKey = /** @class */ (function () {
    function CsdlKey(propertyRefs) {
        this.propertyRefs = propertyRefs;
    }
    return CsdlKey;
}());
export { CsdlKey };
if (false) {
    /** @type {?} */
    CsdlKey.prototype.propertyRefs;
}
var CsdlPropertyRef = /** @class */ (function () {
    function CsdlPropertyRef(name, alias) {
        this.name = name;
        this.alias = alias;
    }
    return CsdlPropertyRef;
}());
export { CsdlPropertyRef };
if (false) {
    /** @type {?} */
    CsdlPropertyRef.prototype.name;
    /** @type {?} */
    CsdlPropertyRef.prototype.alias;
}
