// odata-v4-ng/esm2015/lib/odata-service/http-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function HttpOptionsI() { }
/** @type {?|undefined} */
HttpOptionsI.prototype.headers;
/** @type {?|undefined} */
HttpOptionsI.prototype.params;
/** @type {?|undefined} */
HttpOptionsI.prototype.reportProgress;
/** @type {?|undefined} */
HttpOptionsI.prototype.withCredentials;
export class HttpOptions {
    /**
     * @param {?=} headers
     * @param {?=} observe
     * @param {?=} params
     * @param {?=} reportProgress
     * @param {?=} responseType
     * @param {?=} withCredentials
     */
    constructor(headers, observe = 'response', params, reportProgress, responseType = 'text', withCredentials) {
        this.headers = headers;
        this.observe = observe;
        this.params = params;
        this.reportProgress = reportProgress;
        this.responseType = responseType;
        this.withCredentials = withCredentials;
    }
}
if (false) {
    /** @type {?} */
    HttpOptions.prototype.headers;
    /** @type {?} */
    HttpOptions.prototype.observe;
    /** @type {?} */
    HttpOptions.prototype.params;
    /** @type {?} */
    HttpOptions.prototype.reportProgress;
    /** @type {?} */
    HttpOptions.prototype.responseType;
    /** @type {?} */
    HttpOptions.prototype.withCredentials;
}
