// odata-v4-ng/esm2015/lib/odata-response/odata-response-abstract.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Utils } from '../utils/utils';
/**
 * @abstract
 */
export class ODataResponseAbstract {
    /**
     * @param {?} httpResponse
     */
    constructor(httpResponse) {
        this.httpResponse = httpResponse;
    }
    /**
     * @return {?}
     */
    getHttpResponse() {
        return this.httpResponse;
    }
    /**
     * @return {?}
     */
    isOk() {
        return this.httpResponse.ok;
    }
    /**
     * @return {?}
     */
    getBodyAsJson() {
        return null;
    }
    /**
     * @return {?}
     */
    getBodyAsText() {
        return this.httpResponse.body;
    }
    /**
     * @return {?}
     */
    toString() {
        /** @type {?} */
        let res = `${this.httpResponse.status} ${this.httpResponse.statusText}\n`;
        /** @type {?} */
        const headers = this.httpResponse.headers;
        for (const key of headers.keys()) {
            res += key + ': ';
            /** @type {?} */
            let valueString = '';
            for (const value of headers.getAll(key)) {
                if (valueString.length) {
                    valueString += ' ';
                }
                valueString += value;
            }
            res += valueString + '\n';
        }
        /** @type {?} */
        const json = this.getBodyAsJson();
        if (Utils.isNotNullNorUndefined(json)) {
            res += JSON.stringify(json, null, 4);
        }
        else {
            res += this.getBodyAsText();
        }
        return res;
    }
}
if (false) {
    /** @type {?} */
    ODataResponseAbstract.prototype.httpResponse;
}
