// odata-v4-ng/esm5/lib/odata-service/http-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function HttpOptionsI() { }
/** @type {?|undefined} */
HttpOptionsI.prototype.headers;
/** @type {?|undefined} */
HttpOptionsI.prototype.params;
/** @type {?|undefined} */
HttpOptionsI.prototype.reportProgress;
/** @type {?|undefined} */
HttpOptionsI.prototype.withCredentials;
var HttpOptions = /** @class */ (function () {
    function HttpOptions(headers, observe, params, reportProgress, responseType, withCredentials) {
        if (observe === void 0) { observe = 'response'; }
        if (responseType === void 0) { responseType = 'text'; }
        this.headers = headers;
        this.observe = observe;
        this.params = params;
        this.reportProgress = reportProgress;
        this.responseType = responseType;
        this.withCredentials = withCredentials;
    }
    return HttpOptions;
}());
export { HttpOptions };
if (false) {
    /** @type {?} */
    HttpOptions.prototype.headers;
    /** @type {?} */
    HttpOptions.prototype.observe;
    /** @type {?} */
    HttpOptions.prototype.params;
    /** @type {?} */
    HttpOptions.prototype.reportProgress;
    /** @type {?} */
    HttpOptions.prototype.responseType;
    /** @type {?} */
    HttpOptions.prototype.withCredentials;
}
