// odata-v4-ng/esm5/lib/odata-service/odata.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ODataResponse } from '../odata-response/odata-response';
import { Utils } from '../utils/utils';
import { HttpOptions } from './http-options';
var ODataService = /** @class */ (function () {
    function ODataService(http) {
        this.http = http;
    }
    /**
     * @param {?} odataQuery
     * @param {?=} httpOptions
     * @return {?}
     */
    ODataService.prototype.get = /**
     * @param {?} odataQuery
     * @param {?=} httpOptions
     * @return {?}
     */
    function (odataQuery, httpOptions) {
        /** @type {?} */
        var url = odataQuery.toString();
        /** @type {?} */
        var options = this.createHttpOptions(httpOptions);
        return this.http.get(url, options).pipe(map(function (response) { return new ODataResponse(response); }));
    };
    /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} httpOptions
     * @return {?}
     */
    ODataService.prototype.post = /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} httpOptions
     * @return {?}
     */
    function (odataQuery, body, httpOptions) {
        /** @type {?} */
        var url = odataQuery.toString();
        /** @type {?} */
        var options = this.createHttpOptions(httpOptions);
        return this.http.post(url, body, options).pipe(map(function (response) { return new ODataResponse(response); }));
    };
    /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    ODataService.prototype.patch = /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    function (odataQuery, body, etag, httpOptions) {
        /** @type {?} */
        var url = odataQuery.toString();
        /** @type {?} */
        var options = this.createHttpOptions(httpOptions);
        options = this.mergeETag(options, etag);
        return this.http.patch(url, body, options).pipe(map(function (response) { return new ODataResponse(response); }));
    };
    /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    ODataService.prototype.put = /**
     * @param {?} odataQuery
     * @param {?} body
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    function (odataQuery, body, etag, httpOptions) {
        /** @type {?} */
        var url = odataQuery.toString();
        /** @type {?} */
        var options = this.createHttpOptions(httpOptions);
        options = this.mergeETag(options, etag);
        return this.http.put(url, body, options).pipe(map(function (response) { return new ODataResponse(response); }));
    };
    /**
     * @param {?} odataQuery
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    ODataService.prototype.delete = /**
     * @param {?} odataQuery
     * @param {?=} etag
     * @param {?=} httpOptions
     * @return {?}
     */
    function (odataQuery, etag, httpOptions) {
        /** @type {?} */
        var url = odataQuery.toString();
        /** @type {?} */
        var options = this.createHttpOptions(httpOptions);
        options = this.mergeETag(options, etag);
        return this.http.delete(url, options).pipe(map(function (response) { return new ODataResponse(response); }));
    };
    /**
     * @param {?} httpOptions
     * @return {?}
     */
    ODataService.prototype.createHttpOptions = /**
     * @param {?} httpOptions
     * @return {?}
     */
    function (httpOptions) {
        if (httpOptions instanceof HttpOptions) {
            return httpOptions;
        }
        return Object.assign(new HttpOptions(), httpOptions);
    };
    /**
     * @param {?} httpOptions
     * @param {?} etag
     * @return {?}
     */
    ODataService.prototype.mergeETag = /**
     * @param {?} httpOptions
     * @param {?} etag
     * @return {?}
     */
    function (httpOptions, etag) {
        if (Utils.isNullOrUndefined(etag)) {
            return httpOptions;
        }
        if (Utils.isNullOrUndefined(httpOptions)) {
            httpOptions = new HttpOptions();
        }
        if (Utils.isNullOrUndefined(httpOptions.headers)) {
            httpOptions.headers = new HttpHeaders();
        }
        httpOptions.headers[ODataService.IF_MATCH_HEADER] = etag;
        return httpOptions;
    };
    ODataService.IF_MATCH_HEADER = 'If-Match';
    ODataService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ODataService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return ODataService;
}());
export { ODataService };
if (false) {
    /** @type {?} */
    ODataService.IF_MATCH_HEADER;
    /** @type {?} */
    ODataService.prototype.http;
}
