// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-structured-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export class CsdlStructuredType {
    /**
     * @param {?} name
     * @param {?=} properties
     * @param {?=} navigationProperties
     * @param {?=} baseType
     * @param {?=} openType
     * @param {?=} abstract
     */
    constructor(name, properties, navigationProperties, baseType, openType, abstract) {
        this.name = name;
        this.properties = properties;
        this.navigationProperties = navigationProperties;
        this.baseType = baseType;
        this.openType = openType;
        this.abstract = abstract;
    }
}
if (false) {
    /** @type {?} */
    CsdlStructuredType.prototype.name;
    /** @type {?} */
    CsdlStructuredType.prototype.properties;
    /** @type {?} */
    CsdlStructuredType.prototype.navigationProperties;
    /** @type {?} */
    CsdlStructuredType.prototype.baseType;
    /** @type {?} */
    CsdlStructuredType.prototype.openType;
    /** @type {?} */
    CsdlStructuredType.prototype.abstract;
}
export class CsdlComplexType extends CsdlStructuredType {
    /**
     * @param {?} name
     * @param {?=} properties
     * @param {?=} navigationProperties
     * @param {?=} baseType
     * @param {?=} openType
     * @param {?=} abstract
     */
    constructor(name, properties, navigationProperties, baseType, openType, abstract) {
        super(name, properties, navigationProperties, baseType, openType, abstract);
    }
}
export class CsdlEntityType extends CsdlStructuredType {
    /**
     * @param {?} name
     * @param {?=} key
     * @param {?=} properties
     * @param {?=} navigationProperties
     * @param {?=} baseType
     * @param {?=} openType
     * @param {?=} abstract
     * @param {?=} hasStream
     */
    constructor(name, key, properties, navigationProperties, baseType, openType, abstract, hasStream) {
        super(name, properties, navigationProperties, baseType, openType, abstract);
        this.key = key;
        this.hasStream = hasStream;
    }
}
if (false) {
    /** @type {?} */
    CsdlEntityType.prototype.key;
    /** @type {?} */
    CsdlEntityType.prototype.hasStream;
}
export class CsdlKey {
    /**
     * @param {?} propertyRefs
     */
    constructor(propertyRefs) {
        this.propertyRefs = propertyRefs;
    }
}
if (false) {
    /** @type {?} */
    CsdlKey.prototype.propertyRefs;
}
export class CsdlPropertyRef {
    /**
     * @param {?} name
     * @param {?=} alias
     */
    constructor(name, alias) {
        this.name = name;
        this.alias = alias;
    }
}
if (false) {
    /** @type {?} */
    CsdlPropertyRef.prototype.name;
    /** @type {?} */
    CsdlPropertyRef.prototype.alias;
}
