// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-function-action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export class CsdlFunction {
    /**
     * @param {?} name
     * @param {?} returnType
     * @param {?=} isBound
     * @param {?=} entitySetPath
     * @param {?=} isComposable
     * @param {?=} parameters
     */
    constructor(name, returnType, isBound, entitySetPath, isComposable, parameters) {
        this.name = name;
        this.returnType = returnType;
        this.isBound = isBound;
        this.entitySetPath = entitySetPath;
        this.isComposable = isComposable;
        this.parameters = parameters;
    }
}
if (false) {
    /** @type {?} */
    CsdlFunction.prototype.name;
    /** @type {?} */
    CsdlFunction.prototype.returnType;
    /** @type {?} */
    CsdlFunction.prototype.isBound;
    /** @type {?} */
    CsdlFunction.prototype.entitySetPath;
    /** @type {?} */
    CsdlFunction.prototype.isComposable;
    /** @type {?} */
    CsdlFunction.prototype.parameters;
}
export class CsdlAction {
    /**
     * @param {?} name
     * @param {?=} returnType
     * @param {?=} isBound
     * @param {?=} entitySetPath
     * @param {?=} parameters
     */
    constructor(name, returnType, isBound, entitySetPath, parameters) {
        this.name = name;
        this.returnType = returnType;
        this.isBound = isBound;
        this.entitySetPath = entitySetPath;
        this.parameters = parameters;
    }
}
if (false) {
    /** @type {?} */
    CsdlAction.prototype.name;
    /** @type {?} */
    CsdlAction.prototype.returnType;
    /** @type {?} */
    CsdlAction.prototype.isBound;
    /** @type {?} */
    CsdlAction.prototype.entitySetPath;
    /** @type {?} */
    CsdlAction.prototype.parameters;
}
export class CsdlFunctionImport {
    /**
     * @param {?} name
     * @param {?} functionName
     * @param {?=} entitySet
     * @param {?=} IncludeInServiceDocument
     */
    constructor(name, functionName, entitySet, IncludeInServiceDocument) {
        this.name = name;
        this.functionName = functionName;
        this.entitySet = entitySet;
        this.IncludeInServiceDocument = IncludeInServiceDocument;
    }
}
if (false) {
    /** @type {?} */
    CsdlFunctionImport.prototype.name;
    /** @type {?} */
    CsdlFunctionImport.prototype.functionName;
    /** @type {?} */
    CsdlFunctionImport.prototype.entitySet;
    /** @type {?} */
    CsdlFunctionImport.prototype.IncludeInServiceDocument;
}
export class CsdlActionImport {
    /**
     * @param {?} name
     * @param {?} action
     * @param {?=} entitySet
     */
    constructor(name, action, entitySet) {
        this.name = name;
        this.action = action;
        this.entitySet = entitySet;
    }
}
if (false) {
    /** @type {?} */
    CsdlActionImport.prototype.name;
    /** @type {?} */
    CsdlActionImport.prototype.action;
    /** @type {?} */
    CsdlActionImport.prototype.entitySet;
}
export class CsdlParameter {
    /**
     * @param {?} name
     * @param {?} type
     * @param {?=} nullable
     * @param {?=} maxLength
     * @param {?=} precision
     * @param {?=} scale
     * @param {?=} srid
     */
    constructor(name, type, nullable, maxLength, precision, scale, srid) {
        this.name = name;
        this.type = type;
        this.nullable = nullable;
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.srid = srid;
    }
}
if (false) {
    /** @type {?} */
    CsdlParameter.prototype.name;
    /** @type {?} */
    CsdlParameter.prototype.type;
    /** @type {?} */
    CsdlParameter.prototype.nullable;
    /** @type {?} */
    CsdlParameter.prototype.maxLength;
    /** @type {?} */
    CsdlParameter.prototype.precision;
    /** @type {?} */
    CsdlParameter.prototype.scale;
    /** @type {?} */
    CsdlParameter.prototype.srid;
}
export class CsdlReturnType {
    /**
     * @param {?} type
     * @param {?=} nullable
     * @param {?=} maxLength
     * @param {?=} precision
     * @param {?=} scale
     * @param {?=} srid
     */
    constructor(type, nullable, maxLength, precision, scale, srid) {
        this.type = type;
        this.nullable = nullable;
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.srid = srid;
    }
}
if (false) {
    /** @type {?} */
    CsdlReturnType.prototype.type;
    /** @type {?} */
    CsdlReturnType.prototype.nullable;
    /** @type {?} */
    CsdlReturnType.prototype.maxLength;
    /** @type {?} */
    CsdlReturnType.prototype.precision;
    /** @type {?} */
    CsdlReturnType.prototype.scale;
    /** @type {?} */
    CsdlReturnType.prototype.srid;
}
