// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-enum-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export class CsdlEnumType {
    /**
     * @param {?} name
     * @param {?} members
     * @param {?=} underlyingType
     * @param {?=} isFlags
     */
    constructor(name, members, underlyingType, isFlags) {
        this.name = name;
        this.members = members;
        this.underlyingType = underlyingType;
        this.isFlags = isFlags;
    }
}
if (false) {
    /** @type {?} */
    CsdlEnumType.prototype.name;
    /** @type {?} */
    CsdlEnumType.prototype.members;
    /** @type {?} */
    CsdlEnumType.prototype.underlyingType;
    /** @type {?} */
    CsdlEnumType.prototype.isFlags;
}
export class CsdlEnumMember {
    /**
     * @param {?} name
     * @param {?=} value
     */
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
}
if (false) {
    /** @type {?} */
    CsdlEnumMember.prototype.name;
    /** @type {?} */
    CsdlEnumMember.prototype.value;
}
