// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-type-definition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CsdlAnnotable } from './csdl-annotation';
export class CsdlTypeDefinition extends CsdlAnnotable {
    /**
     * @param {?} name
     * @param {?} underlayingType
     * @param {?=} maxLength
     * @param {?=} precision
     * @param {?=} scale
     * @param {?=} unicode
     * @param {?=} srid
     * @param {?=} annotationList
     */
    constructor(name, underlayingType, maxLength, precision, scale, unicode, srid, annotationList) {
        super(annotationList);
        this.name = name;
        this.underlayingType = underlayingType;
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.unicode = unicode;
        this.srid = srid;
    }
}
if (false) {
    /** @type {?} */
    CsdlTypeDefinition.prototype.name;
    /** @type {?} */
    CsdlTypeDefinition.prototype.underlayingType;
    /** @type {?} */
    CsdlTypeDefinition.prototype.maxLength;
    /** @type {?} */
    CsdlTypeDefinition.prototype.precision;
    /** @type {?} */
    CsdlTypeDefinition.prototype.scale;
    /** @type {?} */
    CsdlTypeDefinition.prototype.unicode;
    /** @type {?} */
    CsdlTypeDefinition.prototype.srid;
}
