// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-reference.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CsdlAnnotable } from './csdl-annotation';
export class CsdlReference extends CsdlAnnotable {
    /**
     * @param {?} uri
     * @param {?=} includes
     * @param {?=} includeAnnotations
     * @param {?=} annotationList
     */
    constructor(uri, includes, includeAnnotations, annotationList) {
        super(annotationList);
        this.uri = uri;
        this.includes = includes;
        this.includeAnnotations = includeAnnotations;
    }
}
if (false) {
    /** @type {?} */
    CsdlReference.prototype.uri;
    /** @type {?} */
    CsdlReference.prototype.includes;
    /** @type {?} */
    CsdlReference.prototype.includeAnnotations;
}
export class CsdlInclude {
    /**
     * @param {?} namespace
     * @param {?=} alias
     */
    constructor(namespace, alias) {
        this.namespace = namespace;
        this.alias = alias;
    }
}
if (false) {
    /** @type {?} */
    CsdlInclude.prototype.namespace;
    /** @type {?} */
    CsdlInclude.prototype.alias;
}
export class CsdlIncludeAnnotations {
    /**
     * @param {?} termNamespace
     * @param {?=} qualifier
     * @param {?=} targetNamespace
     */
    constructor(termNamespace, qualifier, targetNamespace) {
        this.termNamespace = termNamespace;
        this.qualifier = qualifier;
        this.targetNamespace = targetNamespace;
    }
}
if (false) {
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.termNamespace;
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.qualifier;
    /** @type {?} */
    CsdlIncludeAnnotations.prototype.targetNamespace;
}
