// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-annotation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export class CsdlAnnotable {
    /**
     * @param {?=} annotationList
     */
    constructor(annotationList) {
        this.annotationList = annotationList;
    }
}
if (false) {
    /** @type {?} */
    CsdlAnnotable.prototype.annotationList;
}
export class CsdlAnnotations extends CsdlAnnotable {
    /**
     * @param {?} target
     * @param {?} annotationList
     * @param {?=} qualifier
     */
    constructor(target, annotationList, qualifier) {
        super(annotationList);
        this.target = target;
        this.qualifier = qualifier;
    }
}
if (false) {
    /** @type {?} */
    CsdlAnnotations.prototype.target;
    /** @type {?} */
    CsdlAnnotations.prototype.qualifier;
}
export class CsdlAnnotation {
    /**
     * @param {?} term
     * @param {?=} qualifier
     */
    constructor(term, qualifier) {
        this.term = term;
        this.qualifier = qualifier;
    }
}
if (false) {
    /** @type {?} */
    CsdlAnnotation.prototype.term;
    /** @type {?} */
    CsdlAnnotation.prototype.qualifier;
}
export class CsdlTerm {
    /**
     * @param {?} name
     * @param {?} type
     * @param {?=} baseTerm
     * @param {?=} defaultValue
     * @param {?=} appliesTo
     * @param {?=} nullable
     * @param {?=} maxLength
     * @param {?=} precision
     * @param {?=} scale
     * @param {?=} srid
     */
    constructor(name, type, baseTerm, defaultValue, appliesTo, nullable, maxLength, precision, scale, srid) {
        this.name = name;
        this.type = type;
        this.baseTerm = baseTerm;
        this.defaultValue = defaultValue;
        this.appliesTo = appliesTo;
        this.nullable = nullable;
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.srid = srid;
    }
}
if (false) {
    /** @type {?} */
    CsdlTerm.prototype.name;
    /** @type {?} */
    CsdlTerm.prototype.type;
    /** @type {?} */
    CsdlTerm.prototype.baseTerm;
    /** @type {?} */
    CsdlTerm.prototype.defaultValue;
    /** @type {?} */
    CsdlTerm.prototype.appliesTo;
    /** @type {?} */
    CsdlTerm.prototype.nullable;
    /** @type {?} */
    CsdlTerm.prototype.maxLength;
    /** @type {?} */
    CsdlTerm.prototype.precision;
    /** @type {?} */
    CsdlTerm.prototype.scale;
    /** @type {?} */
    CsdlTerm.prototype.srid;
}
