// odata-v4-ng/esm2015/lib/odata-response/csdl/csdl-structural-property.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class CsdlStructuralProperty {
    /**
     * @param {?} name
     * @param {?} type
     * @param {?=} nullable
     */
    constructor(name, type, nullable) {
        this.name = name;
        this.type = type;
        this.nullable = nullable;
    }
}
if (false) {
    /** @type {?} */
    CsdlStructuralProperty.prototype.name;
    /** @type {?} */
    CsdlStructuralProperty.prototype.type;
    /** @type {?} */
    CsdlStructuralProperty.prototype.nullable;
}
export class CsdlProperty extends CsdlStructuralProperty {
    /**
     * @param {?} name
     * @param {?} type
     * @param {?=} nullable
     * @param {?=} maxLength
     * @param {?=} precision
     * @param {?=} scale
     * @param {?=} unicode
     * @param {?=} srid
     * @param {?=} defaultValue
     */
    constructor(name, type, nullable, maxLength, precision, scale, unicode, srid, defaultValue) {
        super(name, type, nullable);
        this.maxLength = maxLength;
        this.precision = precision;
        this.scale = scale;
        this.unicode = unicode;
        this.srid = srid;
        this.defaultValue = defaultValue;
    }
}
if (false) {
    /** @type {?} */
    CsdlProperty.prototype.maxLength;
    /** @type {?} */
    CsdlProperty.prototype.precision;
    /** @type {?} */
    CsdlProperty.prototype.scale;
    /** @type {?} */
    CsdlProperty.prototype.unicode;
    /** @type {?} */
    CsdlProperty.prototype.srid;
    /** @type {?} */
    CsdlProperty.prototype.defaultValue;
}
export class CsdlNavigationProperty extends CsdlStructuralProperty {
    /**
     * @param {?} name
     * @param {?} type
     * @param {?=} nullable
     * @param {?=} partner
     * @param {?=} containsTarget
     * @param {?=} referentialConstraints
     * @param {?=} onDelete
     */
    constructor(name, type, nullable, partner, containsTarget, referentialConstraints, onDelete) {
        super(name, type, nullable);
        this.partner = partner;
        this.containsTarget = containsTarget;
        this.referentialConstraints = referentialConstraints;
        this.onDelete = onDelete;
    }
}
if (false) {
    /** @type {?} */
    CsdlNavigationProperty.prototype.partner;
    /** @type {?} */
    CsdlNavigationProperty.prototype.containsTarget;
    /** @type {?} */
    CsdlNavigationProperty.prototype.referentialConstraints;
    /** @type {?} */
    CsdlNavigationProperty.prototype.onDelete;
}
export class CsdlReferentialConstraint {
    /**
     * @param {?} property
     * @param {?} referencedProperty
     */
    constructor(property, referencedProperty) {
        this.property = property;
        this.referencedProperty = referencedProperty;
    }
}
if (false) {
    /** @type {?} */
    CsdlReferentialConstraint.prototype.property;
    /** @type {?} */
    CsdlReferentialConstraint.prototype.referencedProperty;
}
export class CsdlOnDelete {
    /**
     * @param {?} action
     */
    constructor(action) {
        this.action = action;
    }
}
if (false) {
    /** @type {?} */
    CsdlOnDelete.prototype.action;
}
