// odata-v4-ng/esm5/lib/query-options/expand.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { QueryOptions } from './query-options';
import { Utils } from '../utils/utils';
var Expand = /** @class */ (function () {
    function Expand(entitySet) {
        Utils.requireNotNullNorUndefined(entitySet, 'entitySet');
        Utils.requireNotEmpty(entitySet, 'entitySet');
        this.entitySet = entitySet;
        this.queryOptions = new QueryOptions(Expand.SEPARATOR);
    }
    // QUERY OPTIONS
    /**
     * @param {?} select
     * @return {?}
     */
    Expand.prototype.select = /**
     * @param {?} select
     * @return {?}
     */
    function (select) {
        this.queryOptions.select(select);
        return this;
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    Expand.prototype.filter = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        this.queryOptions.filter(filter);
        return this;
    };
    /**
     * @param {?} expand
     * @return {?}
     */
    Expand.prototype.expand = /**
     * @param {?} expand
     * @return {?}
     */
    function (expand) {
        this.queryOptions.expand(expand);
        return this;
    };
    /**
     * @param {?} orderby
     * @return {?}
     */
    Expand.prototype.orderby = /**
     * @param {?} orderby
     * @return {?}
     */
    function (orderby) {
        this.queryOptions.orderby(orderby);
        return this;
    };
    /**
     * @param {?} search
     * @return {?}
     */
    Expand.prototype.search = /**
     * @param {?} search
     * @return {?}
     */
    function (search) {
        this.queryOptions.search(search);
        return this;
    };
    /**
     * @param {?} skip
     * @return {?}
     */
    Expand.prototype.skip = /**
     * @param {?} skip
     * @return {?}
     */
    function (skip) {
        this.queryOptions.skip(skip);
        return this;
    };
    /**
     * @param {?} top
     * @return {?}
     */
    Expand.prototype.top = /**
     * @param {?} top
     * @return {?}
     */
    function (top) {
        this.queryOptions.top(top);
        return this;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    Expand.prototype.customOption = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.queryOptions.customOption(key, value);
        return this;
    };
    /**
     * @return {?}
     */
    Expand.prototype.toString = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var res = this.entitySet;
        if (!Utils.isEmpty(this.queryOptions)) {
            res += '(' + this.queryOptions + ')';
        }
        return res;
    };
    /**
     * @return {?}
     */
    Expand.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return Utils.isEmpty(this.entitySet);
    };
    Expand.SEPARATOR = ';';
    return Expand;
}());
export { Expand };
if (false) {
    /** @type {?} */
    Expand.SEPARATOR;
    /** @type {?} */
    Expand.prototype.entitySet;
    /** @type {?} */
    Expand.prototype.queryOptions;
}
