// odata-v4-ng/esm2015/lib/query-options/expand.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { QueryOptions } from './query-options';
import { Utils } from '../utils/utils';
export class Expand {
    /**
     * @param {?} entitySet
     */
    constructor(entitySet) {
        Utils.requireNotNullNorUndefined(entitySet, 'entitySet');
        Utils.requireNotEmpty(entitySet, 'entitySet');
        this.entitySet = entitySet;
        this.queryOptions = new QueryOptions(Expand.SEPARATOR);
    }
    /**
     * @param {?} select
     * @return {?}
     */
    select(select) {
        this.queryOptions.select(select);
        return this;
    }
    /**
     * @param {?} filter
     * @return {?}
     */
    filter(filter) {
        this.queryOptions.filter(filter);
        return this;
    }
    /**
     * @param {?} expand
     * @return {?}
     */
    expand(expand) {
        this.queryOptions.expand(expand);
        return this;
    }
    /**
     * @param {?} orderby
     * @return {?}
     */
    orderby(orderby) {
        this.queryOptions.orderby(orderby);
        return this;
    }
    /**
     * @param {?} search
     * @return {?}
     */
    search(search) {
        this.queryOptions.search(search);
        return this;
    }
    /**
     * @param {?} skip
     * @return {?}
     */
    skip(skip) {
        this.queryOptions.skip(skip);
        return this;
    }
    /**
     * @param {?} top
     * @return {?}
     */
    top(top) {
        this.queryOptions.top(top);
        return this;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    customOption(key, value) {
        this.queryOptions.customOption(key, value);
        return this;
    }
    /**
     * @return {?}
     */
    toString() {
        /** @type {?} */
        let res = this.entitySet;
        if (!Utils.isEmpty(this.queryOptions)) {
            res += '(' + this.queryOptions + ')';
        }
        return res;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return Utils.isEmpty(this.entitySet);
    }
}
Expand.SEPARATOR = ';';
if (false) {
    /** @type {?} */
    Expand.SEPARATOR;
    /** @type {?} */
    Expand.prototype.entitySet;
    /** @type {?} */
    Expand.prototype.queryOptions;
}
