// odata-v4-ng/esm2015/lib/query-options/operator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @enum {number} */
const OperatorComparison = {
    EQ: 0, NE: 1, GT: 2, GE: 3, LT: 4, LE: 5, HAS: 6,
};
export { OperatorComparison };
OperatorComparison[OperatorComparison.EQ] = 'EQ';
OperatorComparison[OperatorComparison.NE] = 'NE';
OperatorComparison[OperatorComparison.GT] = 'GT';
OperatorComparison[OperatorComparison.GE] = 'GE';
OperatorComparison[OperatorComparison.LT] = 'LT';
OperatorComparison[OperatorComparison.LE] = 'LE';
OperatorComparison[OperatorComparison.HAS] = 'HAS';
/** @enum {number} */
const OperatorLogical = {
    AND: 0, OR: 1, NOT: 2,
};
export { OperatorLogical };
OperatorLogical[OperatorLogical.AND] = 'AND';
OperatorLogical[OperatorLogical.OR] = 'OR';
OperatorLogical[OperatorLogical.NOT] = 'NOT';
/** @enum {number} */
const OperatorArithmetic = {
    ADD: 0, SUB: 1, MUL: 2, DIV: 3, MOD: 4,
};
export { OperatorArithmetic };
OperatorArithmetic[OperatorArithmetic.ADD] = 'ADD';
OperatorArithmetic[OperatorArithmetic.SUB] = 'SUB';
OperatorArithmetic[OperatorArithmetic.MUL] = 'MUL';
OperatorArithmetic[OperatorArithmetic.DIV] = 'DIV';
OperatorArithmetic[OperatorArithmetic.MOD] = 'MOD';
