// odata-v4-ng/esm5/lib/odata-response/odata-response-abstract.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Utils } from '../utils/utils';
/**
 * @abstract
 */
var /**
 * @abstract
 */
ODataResponseAbstract = /** @class */ (function () {
    function ODataResponseAbstract(httpResponse) {
        this.httpResponse = httpResponse;
    }
    /**
     * @return {?}
     */
    ODataResponseAbstract.prototype.getHttpResponse = /**
     * @return {?}
     */
    function () {
        return this.httpResponse;
    };
    /**
     * @return {?}
     */
    ODataResponseAbstract.prototype.isOk = /**
     * @return {?}
     */
    function () {
        return this.httpResponse.ok;
    };
    /**
     * @return {?}
     */
    ODataResponseAbstract.prototype.getBodyAsJson = /**
     * @return {?}
     */
    function () {
        return null;
    };
    /**
     * @return {?}
     */
    ODataResponseAbstract.prototype.getBodyAsText = /**
     * @return {?}
     */
    function () {
        return this.httpResponse.body;
    };
    /**
     * @return {?}
     */
    ODataResponseAbstract.prototype.toString = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var res = this.httpResponse.status + " " + this.httpResponse.statusText + "\n";
        /** @type {?} */
        var headers = this.httpResponse.headers;
        try {
            for (var _a = tslib_1.__values(headers.keys()), _b = _a.next(); !_b.done; _b = _a.next()) {
                var key = _b.value;
                res += key + ': ';
                /** @type {?} */
                var valueString = '';
                try {
                    for (var _c = tslib_1.__values(headers.getAll(key)), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var value = _d.value;
                        if (valueString.length) {
                            valueString += ' ';
                        }
                        valueString += value;
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_e = _c.return)) _e.call(_c);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                res += valueString + '\n';
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_f = _a.return)) _f.call(_a);
            }
            finally { if (e_2) throw e_2.error; }
        }
        /** @type {?} */
        var json = this.getBodyAsJson();
        if (Utils.isNotNullNorUndefined(json)) {
            res += JSON.stringify(json, null, 4);
        }
        else {
            res += this.getBodyAsText();
        }
        return res;
        var e_2, _f, e_1, _e;
    };
    return ODataResponseAbstract;
}());
/**
 * @abstract
 */
export { ODataResponseAbstract };
if (false) {
    /** @type {?} */
    ODataResponseAbstract.prototype.httpResponse;
}
